--- patched/cryptroot.orig	2023-03-31 12:42:39.752554806 +0200
+++ patched/cryptroot	2023-03-31 12:42:39.756554722 +0200
@@ -150,7 +150,7 @@ setup_mapping() {
 
     local count=0 maxtries="${CRYPTTAB_OPTION_tries:-3}" fstype vg rv
     while [ $maxtries -le 0 ] || [ $count -lt $maxtries ]; do
-        if [ -z "${CRYPTTAB_OPTION_keyscript+x}" ] && [ "$CRYPTTAB_KEY" != "none" ]; then
+        if [ -z "${CRYPTTAB_OPTION_keyscript+x}" ] && ( [ -n "${CRYPTTAB_OPTION_tpm2_device}" ] || [ "$CRYPTTAB_KEY" != "none" ] ); then
             # unlock via keyfile
             unlock_mapping "$CRYPTTAB_KEY"
         else
